src/imap_mag/DB.py

```diff
@@ -42,32 +64,96 @@ def __init__(self, db_url=None):
+            try:
+                self.__active_session = session
+                value = func(self, *args, **kwargs)
+                session.commit()
+
+                return value
+            except Exception as e:
+                session.rollback()
+                raise e
+            finally:
+                session.close()
+                self.__active_session = None
+
+        return wrapper
+
+    def __get_active_session(self) -> Session:
+        if self.__active_session is None:
+            raise ValueError(
+                "No active session. Use decorator @__session_manager to create session."
+            )
+
+        return self.__active_session
 
+    @__session_manager
     def insert_files(self, files: list[File]) -> None:
-        session = self.Session()
-        try:
-            for file in files:
-                # check file does not already exist
-                existing_file = (
-                    session.query(File)
-                    .filter_by(name=file.name, path=file.path)
-                    .first()
-                )
-                if existing_file is not None:
-                    continue
+        session = self.__get_active_session()
+        for file in files:
+            # check file does not already exist
+            existing_file = (
+                session.query(File).filter_by(name=file.name, path=file.path).first()
+            )
+            if existing_file is not None:
+                continue
 
-                session.add(file)
+            session.add(file)
 
-            session.commit()
-        except Exception as e:
-            session.rollback()
-            raise e
-        finally:
-            session.close()
+    @__session_manager
+    def get_download_progress_timestamp(self, item_name: str) -> datetime | None:
+        session = self.__get_active_session()
+        download_progress = (
+            session.query(DownloadProgress).filter_by(item_name=item_name).first()
+        )
+        self.update_download_progress(item_name, last_checked_date=datetime.now())
+
+        if download_progress is None:
+            return None
+
+        return download_progress.progress_timestamp
+
+    @__session_manager
+    def update_download_progress(
+        self,
+        item_name: str,
+        *,
+        progress_timestamp: datetime | None = None,
+        last_checked_date: datetime | None = None,
+    ) -> None:
+        session = self.__get_active_session()
+        download_progress = (
+            session.query(DownloadProgress).filter_by(item_name=item_name).first()
+        )
+
+        if download_progress is None:
+            download_progress = DownloadProgress(item_name=item_name)
+
+        if progress_timestamp is not None:
+            logger.info(
+                f"Updating progress timestamp for {item_name} to {progress_timestamp}."
+            )
+            download_progress.progress_timestamp = progress_timestamp
```

Comment:
This logic should be pushed into the download progress entity so that the calling code looks something like this:
```
progress = db.get_download_progress(id)
progress.record_successful_download(timestamp)
db.save(progress)
```
Otherwise you end up in an anti-pattern known as [Anemic domain model](https://en.wikipedia.org/wiki/Anemic_domain_model)